csharp/src/iceboxnet/Server.cs

```diff
@@ -12,15 +13,19 @@ public static async Task<int> Main(string[] args)
         {
             try
             {
-                using var communicator = new Ice.Communicator(ref args);
+                await using var communicator = new Ice.Communicator(ref args);
 
-                Console.CancelKeyPress += async (sender, eventArgs) =>
+                using var cancellationSource = new CancellationTokenSource();
+
+                Console.CancelKeyPress += (sender, eventArgs) =>
                 {
+                    cancellationSource.Cancel();
                     eventArgs.Cancel = true;
-                    await communicator.ShutdownAsync().ConfigureAwait(false);
+                    _ = communicator.ShutdownAsync();
```

Comment:
What if we only canceled the cancellation token source and not shutdown the communicator here? And instead of awaiting on `_communicator.WaitForShutdownAsync()` we could simply await on `Task.Delay(-1, cancel)`. I.e.: wait indefinitely for the cancellation. Once this await returns we would call `await _communicator.ShutdownAsync()`.

Reply:
For me, we regular pattern is to call `communicator.ShutdownAsync` on Ctrl+C, and here as a bonus we use a cancellation token / cancellation source to cancel awaiting tasks to speed up the shutdown.